redis/pool.go

```diff
@@ -299,34 +299,61 @@ func (p *Pool) lazyInit() {
 	p.mu.Unlock()
 }
 
-// get prunes stale connections and returns a connection from the idle list or
-// creates a new connection.
-func (p *Pool) get(ctx context.Context) (*poolConn, error) {
+// waitVacantConn waits while at least one vacant connection appears in the pool.
+// If `ctx` expires before that, an error is returned.
+//
+// waitVacantConn returns:
+// - (as its first return value) the time it spent waiting, if there were no vacant connection in the pool right away
+// or zero otherwise
+// - (as its second return value) error, if any
+func (p *Pool) waitVacantConn(ctx context.Context) (time.Duration, error) {
+	if !p.Wait {
+		return 0, nil
+	}
+	if p.MaxActive <= 0 {
+		return 0, nil
+	}
 
-	// Handle limit for p.Wait == true.
-	var waited time.Duration
-	if p.Wait && p.MaxActive > 0 {
-		p.lazyInit()
-
-		// wait indicates if we believe it will block so its not 100% accurate
-		// however for stats it should be good enough.
-		wait := len(p.ch) == 0
-		var start time.Time
-		if wait {
-			start = time.Now()
-		}
-		if ctx == nil {
-			<-p.ch
-		} else {
+	p.lazyInit()
+
+	// wait indicates if we believe it will block so its not 100% accurate
+	// however for stats it should be good enough.
+	wait := len(p.ch) == 0
+	var start time.Time
+	if wait {
+		start = time.Now()
+	}
+
+	if ctx == nil {
+		<-p.ch
+	} else {
+		select {
+		case <-p.ch:
+			// Additionally check that context hasn't expired while we were waiting,
+			// because `select` picks a random `case` if several of them are "ready".
 			select {
-			case <-p.ch:
```

Comment:
While I don't mind the refactor its arguably simpler to just add the additional select in here and be done:
```go
select {
case <-ctx.Done()
    return nil, ctx.Err()
default:
}
```

Reply:
I've changed some bits (as you've suggested), 

if you still prefer to have the original version of the code with "extra" select, I'll remove refactoring from PR